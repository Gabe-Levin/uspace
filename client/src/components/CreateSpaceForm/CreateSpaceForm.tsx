import React, { useState } from 'react';
import '../CreateEntryForm/CreateEntryForm.scss';
import { TextInput, Textarea } from '@mantine/core';
import DOMPurify from 'dompurify';
import SpaceWithCreatorType, {
  SpaceDataType,
} from '../../interfaces/Interfaces';
import { useAuth0 } from '@auth0/auth0-react';
import API_SERVICE from '../../Api-Service';
import _ from 'lodash';
import { useNavigate } from 'react-router';

interface Incoming {
  setOpened: Function;
  allSpaces: SpaceWithCreatorType[];
  setAllSpaces: Function;
}

function CreateSpaceForm(props: Incoming) {
  const [name, setName] = useState<string>('');
  const [description, setDescription] = useState<string>('');
  const { user } = useAuth0();
  const navigate = useNavigate();

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const spaceData = {
      name: DOMPurify.sanitize(name),
      description: DOMPurify.sanitize(description),
    };

    // create new space
    const newSpace: SpaceDataType = await API_SERVICE.createSpace(spaceData);

    // check if user exists
    if (user) {
      // fetch user from db to get id
      const foundUser = await API_SERVICE.findUserBySub(user.sub!);
      // create a m-m relationship using the user id, space id and 2 for creator
      await API_SERVICE.createUserSpaceRole(foundUser.id, newSpace.id, 2);
      // add new userSpaceRole array and input new user
      newSpace.User_Space_Role = [
        {
          user: {
            email: foundUser.email,
            username: foundUser.username,
            picture_url: foundUser.picture_url,
          },
        },
      ];

      // overwrite posts state with the cloned posts incl. the new comment
      props.setAllSpaces(cloneSpacesAndAddNewSpace(props.allSpaces, newSpace));

      // direct to new space
      navigate(`/spaces/${newSpace.id}`); //TODO: add this back
    }

    // hide modal
    props.setOpened(false);
  };

  // deep clone given space and add new space
  const cloneSpacesAndAddNewSpace = (
    spaces: SpaceWithCreatorType[],
    newSpace: SpaceDataType
  ) => {
    // deep clone given spaces
    const clonedSpaces = _.cloneDeep(spaces);
    // push space to cloned spaces
    clonedSpaces.push(newSpace);
    return clonedSpaces;
  };

  return (
    <div className="create-entry-form">
      <form onSubmit={handleSubmit}>
        <label>Space name:</label>
        <TextInput
          required
          value={name}
          onChange={(event) => setName(event.currentTarget.value)}
        />
        <label>Description:</label>
        <Textarea
          required
          value={description}
          onChange={(event) => setDescription(event.currentTarget.value)}
        />
        <button className="create-entry-form-submit">Create</button>
      </form>
    </div>
  );
}

export default CreateSpaceForm;
