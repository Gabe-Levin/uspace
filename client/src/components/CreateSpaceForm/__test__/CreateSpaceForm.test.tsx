import React from 'react';
import CreateSpaceForm from '../CreateSpaceForm';
import { render, fireEvent, cleanup } from '@testing-library/react';
import { screen } from '@testing-library/dom';
import '@testing-library/jest-dom/extend-expect';
import userEvent from '@testing-library/user-event';
import renderer from 'react-test-renderer';

const mockDate = new Date();
const mockedUseNavigate = jest.fn();
const propsMock = {
  setOpened: jest.fn(),
  allSpaces: [
    {id: 111, name: 'space 1 name', createdAt: mockDate, userSpaceRoles: [
      {user:{username:'billy', email: 'billy@billy.com'}},{user: {username:'sammy', email: 'sammy@gsammy.com'}}
    ]},
    {id: 222, name: 'space 2 name', createdAt: mockDate, userSpaceRoles: [
      {user: {username:'ms', email: 'ms@ms.com'}}, {user: {username:'mavi', email: 'mavi@gmavi.com'}}
    ]}
  ],
  setAllSpaces: jest.fn()
}
jest.mock('react-router', () => ({
  ...jest.requireActual('react-router') as any,
  useNavigate: () => mockedUseNavigate
}));

describe ('Testing CreateSpaceFrom component', () => {
  afterEach(cleanup);
  test('contains a field to enter space name', async () => {
    render(<CreateSpaceForm setOpened={propsMock.setOpened} allSpaces={propsMock.allSpaces} setAllSpaces={propsMock.setAllSpaces} />)
    const spaceNameInput = screen.getByLabelText(/space name:/i);
    await userEvent.type(spaceNameInput, 'new space', {delay: 1});
    // @ts-ignore
    expect(spaceNameInput.value).toBe('new space');
  })
  test('contains a field to enter space description', async () => {
    render(<CreateSpaceForm setOpened={propsMock.setOpened} allSpaces={propsMock.allSpaces} setAllSpaces={propsMock.setAllSpaces} />)
    const spaceDescriptionInput = screen.getByLabelText(/description:/i);
    await userEvent.type(spaceDescriptionInput, 'new space description', {delay: 1});
    // @ts-ignore
    expect(spaceDescriptionInput.value).toBe('new space description');
  })
  test('snapshot matches, component renders correctly', () => {
    const tree = renderer
      .create(<CreateSpaceForm setOpened={propsMock.setOpened} allSpaces={propsMock.allSpaces} setAllSpaces={propsMock.setAllSpaces} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  })
  // Not finished
  // test('prevents default on click', async () => {
  //   render(<CreateSpaceForm setOpened={propsMock.setOpened} allSpaces={propsMock.allSpaces} setAllSpaces={propsMock.setAllSpaces} />);
  //   const button = screen.getByRole('button', { name: /create/i });
  //   const buttonClickEvent = new MouseEvent('click');
  //   Object.assign(buttonClickEvent, {preventDefault: jest.fn()});
  //   fireEvent.click(button, buttonClickEvent);
  //   expect(buttonClickEvent.preventDefault).toBeCalled();
  // })
})
