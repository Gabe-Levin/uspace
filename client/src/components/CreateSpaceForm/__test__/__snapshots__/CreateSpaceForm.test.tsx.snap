// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing CreateSpaceFrom component snapshot matches, component renders correctly 1`] = `
<div
  className="create-entry-form"
>
  <form
    onSubmit={[Function]}
  >
    <label
      htmlFor="spaceName"
    >
      Space name:
    </label>
    <div
      className="mantine-TextInput-root mantine-18udhi"
    >
      <div
        className="mantine-TextInput-wrapper mantine-12sbrde"
      >
        <input
          aria-invalid={false}
          className="mantine-TextInput-defaultVariant mantine-TextInput-input mantine-us7vc8"
          id="spaceName"
          name="spaceName"
          onChange={[Function]}
          required={true}
          type="text"
          value=""
        />
      </div>
    </div>
    <label
      htmlFor="spaceDescription"
    >
      Description:
    </label>
    <div
      className="mantine-Textarea-root mantine-18udhi"
    >
      <div
        className="mantine-Textarea-wrapper mantine-12sbrde"
      >
        <textarea
          aria-invalid={false}
          className="mantine-Textarea-defaultVariant mantine-Textarea-input mantine-wcnwz0"
          id="spaceDescription"
          name="spaceDescription"
          onChange={[Function]}
          required={true}
          value=""
        />
      </div>
    </div>
    <button
      className="create-entry-form-submit"
    >
      Create
    </button>
  </form>
</div>
`;
